File: solutions/elixir/darts/6/lib/darts.ex
defmodule Darts do

  def score(impact_coordinates),
    do:
      impact_coordinates
      |> zone()
      |> points()

  defp zone({x, y}), do: ring(:math.sqrt(x**2 + y**2))

  defp ring(radius) when radius <= 1,  do: :inner
  defp ring(radius) when radius <= 5,  do: :middle
  defp ring(radius) when radius <= 10, do: :outer
  defp ring(_),                        do: :outside

  defp points(:inner),   do: 10
  defp points(:middle),  do: 5
  defp points(:outer),   do: 1
  defp points(:outside), do: 0
  
end

File: solutions/haskell/darts/4/src/Darts.hs
module Darts (score) where

data Zone = Inner | Middle | Outer | Outside

score :: Float -> Float -> Int
score x y = points . circle $ radius x y

radius :: Float -> Float -> Float
radius x y = sqrt (x*x + y*y)

circle :: Float -> Zone
circle r
  | r <= 1    = Inner
  | r <= 5    = Middle
  | r <= 10   = Outer
  | otherwise = Outside

points :: Zone -> Int
points Inner   = 10
points Middle  = 5
points Outer   = 1
points Outside = 0
